proof/invariant-abstract/DetSchedInvs_AI.thy

```diff
@@ -916,6 +917,54 @@ global_interpretation ep_recv_qs:
   opt_map_opt_map_cons_def_locale _ ep_of eps_of_kh Endpoint recv_q_of RecvEP ep_recv_qs_of_eps
   using ep_recv_qs_of_eps_def recv_q_of_None recv_q_of_Some by unfold_locales
 
+\<comment> \<open>Notifications\<close>
+
+\<comment> \<open>Project notifications from the kernel heap\<close>
+
+primrec ntfn_of :: "Structures_A.kernel_object \<rightharpoonup> notification" where
+  "ntfn_of (TCB tcb) = None"
+| "ntfn_of (Endpoint ep) = None"
+| "ntfn_of (Notification ntfn) = Some ntfn"
+| "ntfn_of (CNode _ _) = None"
+| "ntfn_of (Reply _) = None"
+| "ntfn_of (SchedContext _ _) = None"
+| "ntfn_of (ArchObj _) = None"
+
+lemma ntfn_of_SomeD[dest!]:
+  "ntfn_of ko = Some obj \<Longrightarrow> ko = Notification obj"
+  by (cases ko; simp)
+
+lemma ntfn_of_NoneD[simp]:
+  "ntfn_of ko = None \<Longrightarrow> \<not> (\<exists>ep. ko = Notification ep)"
+  by (cases ko; simp)
+
+abbreviation ntfns_of :: "'z state \<Rightarrow> obj_ref \<rightharpoonup> Structures_A.notification" where
+  "ntfns_of \<equiv> (\<lambda>s. kheap s |> ntfn_of)"
+
+\<comment> \<open>Replies\<close>
+
+\<comment> \<open>Project replies from the kernel heap\<close>
+
+primrec reply_of :: "Structures_A.kernel_object \<rightharpoonup> reply" where
```

Comment:
Same here (`definition` comment). Just marking all of them so you don't have to search.